### PR TITLE
Use parity to wait for mbarrier

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Today (assuming 4 stages):
```C++
__shared__ uint64_t mbarrier[4];
__shared__ uint32_t tokens[4];
for i in 3:
  tokens[i] = arrive(mbarrier[i]);
  TMA(i);
for i in I:
  tokens[(i + 3) % 4] = arrive(mbarrier[(i + 3) % 4]);
  TMA((i + 3) % 4);
  wait(mbarrier[i % 4], tokens[i % 4]);
  MMA(i);
```

Changed:
```C++
__shared__ uint64_t mbarrier[4];
for i in 3:
  arrive(mbarrier[i]);
  TMA(i);
for i in I:
  arrive(mbarrier[(i + 3) % 4]);
  TMA((i + 3) % 4);
  waitParity(mbarrier[i % 4], (i / 4) % 2);
  MMA(i);
```

Why making this change? The primary reason is to gain more flexibility on moving code around. In the changed code, the thing to wait is only a function of loop index, and is computed inline. It can be moved around by just cut-paste. There is nothing to worry about like "How do I get which token to wait on?", and there is no need to worry about data races on `tokens` when moving things around.

Perf together with https://github.com/NVIDIA/Fuser/pull/3153:

```C++
 Time (%)  Total Time (ns)  Instances  Avg (ns)   Med (ns)   Min (ns)  Max (ns)  StdDev (ns)                                                  Name

 --------  ---------------  ---------  ---------  ---------  --------  --------  -----------  ----------------------------------------------------------------------------------------------------
     20.1          1401712          1  1401712.0  1401712.0   1401712   1401712          0.0  <unnamed>::nvfuser_none_f0_c0_r0_g0(<unnamed>::Tensor<<unnamed>::__half, (int)3, (int)3>, <unnamed>…
     11.2           780952          1   780952.0   780952.0    780952    780952          0.0  nvjet_hsh_192x192_64x3_2x1_v_bz_coopB_NTN
```

Perf nvFuser/cuBLAS: `55.7%`, vs only https://github.com/NVIDIA/Fuser/pull/3153 `52.7%`

The perf diff for the current schedule (64x256 per CTA) is not significant, but if I use 192x192 tile size for each CTA (same as cuBLAS, has larger theoretical throughput), the perf improvement is bigger `1811248 ns` vs `2168390 ns`.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/device_lower/pass/circular_buffer.cpp [hunk 1]
```diff
@@ -689,7 +680,7 @@ class CloneTmaCircularBufferLoopAndInsertSync
 
     // The other threads issue arriveExpectTx without any expected transactions.
     kir::MBarrierArrive* thread_arrive = IrBuilder::create<kir::MBarrierArrive>(
-        mbarrier_arrive_tx_->state(), mbarrier_arrive_tx_->mbarrier());
+        nullptr, mbarrier_arrive_tx_->mbarrier());
```
Comment (rdspring1) [inline csrc/device_lower/pass/circular_buffer.cpp:683]:
```suggestion
        /*state=*/nullptr, mbarrier_arrive_tx_->mbarrier());
```

File: csrc/device_lower/pass/circular_buffer.cpp [hunk 2]
```diff
@@ -753,46 +743,50 @@ class CloneTmaCircularBufferLoopAndInsertSync
...
-        GpuLower::current()->tmaCircularBufferInfo().getMBarrierToken(ldst);
-    kir::TensorIndex* stage_token =
-        IrBuilder::create<kir::TensorIndex>(all_mbarrier_tokens, loop_index);
-
     Val* tx_count = GpuLower::current()->commonScalarMap().hoistScalar(
         getSizeOfTmaLoad(ldst), for_loop_stack_);
     kir::MBarrierArriveExpectTx* mbarrier_arrive_tx =
         IrBuilder::create<kir::MBarrierArriveExpectTx>(
-            stage_token, stage_mbarrier, tx_count);
+            nullptr, stage_mbarrier, tx_count);
```
Comment (rdspring1) [inline csrc/device_lower/pass/circular_buffer.cpp:750]:
```suggestion
            /*state=*/nullptr, stage_mbarrier, tx_count);
```

File: csrc/device_lower/pass/circular_buffer.cpp [hunk 3]
```diff
@@ -753,46 +743,50 @@ class CloneTmaCircularBufferLoopAndInsertSync
...
-    kir::TensorIndex* stage_token =
-        IrBuilder::create<kir::TensorIndex>(all_mbarrier_tokens, loop_index);
-
-    kir::MBarrierWait* mbarrier_wait =
-        IrBuilder::create<kir::MBarrierWait>(stage_mbarrier, stage_token);
+        IrBuilder::create<kir::TensorIndex>(all_mbarriers, stage);
+
+    // Get mbarrier_parity for this circular buffer stage.
+    auto depth = IrBuilder::create<Val>(stage_depth, DataType::UInt32);
+    auto two = IrBuilder::create<Val>(2, DataType::UInt32);
```
Comment (rdspring1) [inline csrc/device_lower/pass/circular_buffer.cpp:775]: Comment about stage parity:

We have an mbarrier for each circular buffer stage, so `loop_index / stage_depth` is `loop_index_per_stage`.
The valid values of phaseParity operand are 0 and 1, so we take the modulo of `loop_index_per_stage` with a divisor of 2. See https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-mbarrier-test-wait-mbarrier-try-wait for reference.

**Review by rdspring1 — APPROVED** [review]
The code looks fine to me assuming all the tma circular buffering tests pass.

**Comment by zasdfgbnm** [comment]
!build

**Pull request merged**